xdsl/dialects/comb.py

```diff
@@ -153,21 +227,115 @@ class XorOp(VariadicCombOperation):
     name = "comb.xor"
 
 
+@dataclass
+class ComparisonOperation:
+    """
+    A generic comparison operation, operation definitions inherit this class.
+
+    The first argument to these comparison operations is the type of comparison
+    being performed, the following comparisons are supported:
+
+    -   equal (mnemonic: `"eq"`; integer value: `0`)
+    -   not equal (mnemonic: `"ne"`; integer value: `1`)
+    -   signed less than (mnemonic: `"slt"`; integer value: `2`)
+    -   signed less than or equal (mnemonic: `"sle"`; integer value: `3`)
+    -   signed greater than (mnemonic: `"sgt"`; integer value: `4`)
+    -   signed greater than or equal (mnemonic: `"sge"`; integer value: `5`)
+    -   unsigned less than (mnemonic: `"ult"`; integer value: `6`)
+    -   unsigned less than or equal (mnemonic: `"ule"`; integer value: `7`)
+    -   unsigned greater than (mnemonic: `"ugt"`; integer value: `8`)
+    -   unsigned greater than or equal (mnemonic: `"uge"`; integer value: `9`)
+    """
+
+    @staticmethod
+    def _get_comparison_predicate(
+        mnemonic: str, comparison_operations: dict[str, int]
+    ) -> int:
+        if mnemonic in comparison_operations:
+            return comparison_operations[mnemonic]
+        else:
+            raise VerifyException(f"Unknown comparison mnemonic: {mnemonic}")
+
+    @staticmethod
+    def _validate_operand_types(operand1: SSAValue, operand2: SSAValue):
+        if operand1.type != operand2.type:
+            raise TypeError(
+                f"Comparison operands must have same type, but "
+                f"provided {operand1.type} and {operand2.type}"
+            )
+
+
 @irdl_op_definition
-class ICmpOp(IRDLOperation):
+class ICmpOp(IRDLOperation, ComparisonOperation):
     """Integer comparison"""
 
     name = "comb.icmp"
 
     T = Annotated[IntegerType, ConstraintVar("T")]
 
+    predicate: IntegerAttr[IndexType] = attr_def(IntegerAttr[IndexType])
     lhs: Operand = operand_def(T)
     rhs: Operand = operand_def(T)
     result: OpResult = result_def(IntegerType(1))
 
-    predicate: IntegerAttr[IndexType] = attr_def(IntegerAttr[IndexType])
     two_state: UnitAttr = attr_def(UnitAttr)
 
+    def __init__(
+        self,
+        operand1: Operation | SSAValue,
+        operand2: Operation | SSAValue,
+        arg: int | str,
+    ):
+        operand1 = SSAValue.get(operand1)
+        operand2 = SSAValue.get(operand2)
+        ICmpOp._validate_operand_types(operand1, operand2)
```

Comment:
This seems off to me, we usually leave all validation to the `validate()` call. This is because there are no guarantees about validity of IR during rewrites.
```suggestion
```

Reply:
I checked `arith` and the same thing happens there, hence I didn't question it, but I think this suggestion actually makes sense.
@math-fehr ?